sdk/core/azure-core/src/main/java/com/azure/core/util/ClientOptions.java

```diff
@@ -0,0 +1,168 @@
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Client Options for setting common properties for example applicationId. Most of these properties are applied on
+ * request being send to Azure Service but some could be used for other purpose also for example applicationId could be
+ * used for telemetry.
+ */
+@Fluent
+public final class ClientOptions {
+    private final ClientLogger logger = new ClientLogger(ClientOptions.class);
+    private final Map<String, Header> headers = new ConcurrentHashMap<>();
+
+    private static final int MAX_APPLICATION_ID_LENGTH = 24;
+    private String applicationId;
+
+    /**
+     * Gets the applicationId.
+     *
+     * @return The applicationId.
+     */
+    public String getApplicationId() {
+        return applicationId;
+    }
+
+    /**
+     * Sets the applicationId provided.
+     *
+     * @param applicationId to be set.
+     * @return updated {@link ClientOptions}.
+     *
+     */
+    public ClientOptions setApplicationId(String applicationId) {
+        if (!CoreUtils.isNullOrEmpty(applicationId)) {
+            if (applicationId.length() > MAX_APPLICATION_ID_LENGTH) {
+                throw logger
+                    .logExceptionAsError(new IllegalArgumentException("'applicationId' length cannot be greater than "
+                        + MAX_APPLICATION_ID_LENGTH));
+            } else if (applicationId.contains(" ")) {
+                throw logger
+                    .logExceptionAsError(new IllegalArgumentException("'applicationId' must not contain a space."));
+            } else {
+                this.applicationId = applicationId;
+            }
+        }
+        return this;
+    }
+
+    /**
+     * Sets the provided headers.
+     *
+     * @param headers headers to be set.
+     * @return updated {@link ClientOptions}.
+     *
+     * @throws NullPointerException if {@code headers} is null.
+     */
+    public ClientOptions headers(Iterable<Header> headers) {
+        Objects.requireNonNull(headers, "'headers' cannot be null.");
+
+        for (final Header header : headers) {
+            this.headers.put(formatKey(header.getName()), header);
+        }
+        return this;
+    }
+
+    /**
+     * Sets a {@link Header header} with the given name and value or append the {@code value} separated by {@code comma}
+     * if the {@link Header} exists for given {@code name}.
+     *
+     * @param name the name
+     * @param value the value
+     * @return The updated ClientOptions object
+     *
+     * @throws NullPointerException if {@code name} or {@code value} is null.
+     */
+    public ClientOptions addHeader(String name, String value) {
+        Objects.requireNonNull(name, "'name' cannot be null.");
+        Objects.requireNonNull(value, "'value' cannot be null.");
+
+        Header existing = get(name);
+        if (existing == null) {
+            this.headers.put(formatKey(name), new Header(name, value));
```

Comment:
If you do need the concurrent hash map, you are not leveraging it by splitting the operations in two.

Reply:
removed